### ISSUE TITLE
extend hostgroup module with non-Foreman options

### ISSUE THREAD

**Issue by evgeni** [theforeman/foreman-ansible-modules#374]
> This is great @bagasse! I've added two comments, but nothing major ;)
> 
> As for OpenSCAP, yeah this sounds like a good addition, I don't mind it being in this PR or an own one. Just remember that we'll need to use a different `apidoc.json` then ;)

@evgeni It would be nice to add extensible smart_proxy support if we can. We use katello and it would be nice to add content proxy also, but maybe we have to make a subclass katello_hostgroup for that ? I didn't wanted to make this happen in this PR that I think  is big enough :)

_Originally posted by @bagasse in https://github.com/theforeman/foreman-ansible-modules/pull/366#issuecomment-514528037_

**Comment by evgeni** [comment]
Braindump: I think we should not create a separate katello_hostgroup module, but just extend the existing one and document that some options won't work if a specific plugin (scap, katello, salt, etc) is not installed.

IIRC we have this in some places already. (Or rather: for some modules behavior on Foreman 1.22 is different to the one on 1.21)

**Comment by mdellweg** [comment]
We _almost_ have it for the snapshot plugin / module. ;)
I also vote for a single module for hostgroups. I know, we could add a second test_playbook especially for those features, but i am unsure whether it were a benefit at all, since one would be forced to rerecord that anyway. All in all, i think we would need a good balance for the amount of setup we want to require to rerecord things. At least it should be documented in a very instructive way.

__Brainstorm:__ We could add a script / provide a playbook to do the needed setup in a dedicated forklift box.

**Comment by bagasse** [comment]
@evgeni, @mdellweg, It's fine for me to document plugin specific params on the `foreman_hostgroup` module I think this module is just missing parameters support for bare foreman server type (regarding to apidoc on foreman 1.22 / develop). If we include katello's specific params and some other plugins like openscap, I think missing params for this module are:
* parameters (foreman, param type is still  missing in develop branch, needs a foreman change)
* openscap_proxy (openscap)
* content_source (katello)
* content_view (katello)
* lifecycle_environment (katello)
* kickstart_repository (katello)

I wanted to tacle this in another PR than #366

**Comment by bagasse** [comment]
On the dev setup part, +1 for some playbooks and and boxes in forklift ! For me, setting up dev env for foreman ansible modules is a barrier to contributions, and we have to make it lower as possible. I already postponed some contributions (like this one, it's been a lot of time that I wanted to contribute on this part because we use it extensively) because a lot of modules have to interact with a lot of foreman-proxies components (DNS, compute resource..). I think we have to provide an all-in-one forklift box that permit to make most of the things and use variables in our test playbooks pointing to this setup by default if possible to ease test recording.

For example my current test env for foreman ansible modules is composed of two boxes (that can be merged in one if DNS smart-proxy is included in foreman server, I don't wanted to fix that in the hostgroup PR because it's unrelated). I used forklift to set it up with following changes:

```
cat playbooks/katello_libvirt.yml 
---
- name: Apply katello playbook
  import_playbook: katello.yml
- name: Apply libvirt playbook
  import_playbook: libvirt.yml
```

```
cat vagrant/settings.yaml
---
memory: 4608
cpus: 2
scale_memory: 1
scale_cpus: 1
sync_type: 'rsync'

cachier:
  mount_options: ['rw', 'vers=3', 'tcp', 'nolock']

ssh_forward_agent: True

domain: 'example.com'
```

```
cat vagrant/boxes.d/99-local.yaml
---
foreman:
    box:   centos7
    memory: 6144
    ansible:
      playbook: 'playbooks/katello_libvirt.yml'
      variables:
        foreman_repositories_version: 1.22
        katello_repositories_version: 3.12
        foreman_installer_additional_packages: ['tfm-rubygem-foreman_dhcp_browser', 'tfm-rubygem-foreman_expire_hosts', 'tfm-rubygem-ovirt_provision_plugin', 'katello']
        foreman_installer_options: ['--enable-foreman-compute-openstack', '--enable-foreman-compute-libvirt', '--enable-foreman-compute-ovirt', '--enable-foreman-plugin-openscap']

# for the DNS part, can be merged with the first box
centos7-devel:
    box: centos7
    memory: 2048
    ansible:
      playbook: 'playbooks/foreman_proxy_content.yml'
      group: 'foreman-proxy-content'
      server: foreman
      variables:
        foreman_repositories_version: 1.22
        katello_repositories_version: 3.12
        foreman_installer_options: ['--foreman-proxy-dns=true']
```

but and AIO box for test should look like

```
foreman:
    box:   centos7
    memory: 6144
    ansible:
      playbook: 'playbooks/katello_libvirt.yml'
      variables:
        foreman_repositories_version: 1.22
        katello_repositories_version: 3.12
        foreman_installer_additional_packages:
          - tfm-rubygem-foreman_dhcp_browser
          - tfm-rubygem-foreman_expire_hosts
          - tfm-rubygem-ovirt_provision_plugin
          - katello
        foreman_installer_options:
          - '--enable-foreman-compute-openstack'
          - '--enable-foreman-compute-libvirt'
          - '--enable-foreman-compute-ovirt'
          - '--enable-foreman-plugin-openscap'
          - '--enable-foreman-proxy-plugin-openscap'
          - '--foreman-proxy-dns=true'
          - '--foreman-proxy-bmc=true'
          - '--foreman-proxy-realm'
```

**Comment by bagasse** [comment]
I'm working on adding missing hostgroup options, but for katello specific stuff as it don't use katello api endpoint, it don't work out of the box:
Organizations created via foreman ansible modules create a 'foreman' org, and not a 'katello' one, so creating content view fail because of missing pieces on candlepin side (404 error when browsing content_views in foreman)
And API don't find any katello resource (like content views)

can't we automaticaly detect if we are talking to forman or katello instance to use the good api endpoints ?

**Comment by evgeni** [comment]
@bagasse we already are (outside of tests). Your changes should work fine "in prod"

What is fooling you (your tests) is that our tests link hostgroup to the foreman apidoc, see https://github.com/theforeman/foreman-ansible-modules/blob/master/tests/fixtures/apidoc/hostgroup.json

If you change that link to `katello.json` stuff will work again. But then we never test the `hostgroup` module against a plain Foreman and @mdellweg will kill me.
Seems we need to think about "one module, many apidocs, many playbooks" sooner than we expected.

**Comment by mdellweg** [comment]
DISCLAMER: I am not as violent as the previous post implies.

I do not see a big problem for running two tests for hostgroup. It is a big module. It deserves it. Also after adding katello-specific properties, it would seem advisable to test, that it still works with plain foreman. So :+1: for having a hostgroup, and a katello_hostgroup test playbook.

**Comment by evgeni** [comment]
Right, and if we call the second playbook `katello_hostgroup.yml`, it will automatically use `katello_hostgroup.json` for the apidoc, so that's solved too. nice.

**Comment by mdellweg** [comment]
My sentiments exactly.

**Comment by bagasse** [comment]
@evgeni thank you for the (very quick) reply ;). As it's not my intention to make shed the blood of anyone, I will move the links of orgs and hostgroup to katello on my local test env to see if all is fine with these changes and then wait for you or @mdellweg instructions :)

**Comment by bagasse** [comment]
@mdellweg, @evgeni thank you both,  I will  follow your advices.

**Comment by bagasse** [comment]
@evgeni, @mdellweg ok I think I've two distinct problems when I try to record the tests:
* For openscap, the openscap_proxy_id is not sent to foreman server and I encounter some an exception in `_flatten_entity` (because when called by `_update_entity` on `current_entity`, the `value` is `None` for the `openscap_proxy` entity so it can't get `value['id']`). I'm wondering if this param is not sent to the server because there is no reference in `tests/fixtures/apidoc/{foreman,katello}.json` or if it's unrelated (and if it's related, maybe it will fix the exception in `_flatten_entity`)
* For katello's params, all katello resources are scoped in an organization (content_views, lifecyle env), so I don't know how to handle that. I see 3 ways, but none are ideal IMHO:
  1. Add an `organization` param to `hostgroup` and use it for all katello's scoped resources (and document that it will only used for that)
  2. Change these params in a dict with organization and resource name instead of only resource name. We have to check that the org if the same for CV and lifcycle env
  3. Add a katello_hostgroup module that inherit foreman_hostgroup and add an organization param and other katello related params


What do you prefer ? Any other way ?

**Comment by mdellweg** [comment]
For the first point, it is probably, that the apidoc was generated without a openscap plugin.

For the second point (given that this is ugly in any case), how is it handled in the UI? Can you select from all CVs or just the ones of the current org?

**Comment by bagasse** [comment]
@mdellweg, for the second point, you can only choose/view resources in the current org, that's why I prefer option 1 and then after 3 (the 2 is very very ugly).

**Comment by mdellweg** [comment]
There is also the option, that you search for those entities in all organizations, that are selected for the host_group.

**Comment by bagasse** [comment]
Yes, but it will not act as the UI, and I want to avoid this.

**Comment by evgeni** [comment]
I think I'd prefer one module for everything if possible. And yes our Foreman modules need a way to be scoped properly when in a katello-like environment (see also https://github.com/theforeman/foreman-ansible-modules/issues/338).

Given that a hostgroup can be org scoped on a plain foreman too, I think it's the best approach to add a hostgroup param.

**Comment by bagasse** [comment]
Ok, it seems that there is missing api endpoint for lifecycle environments so we cannot add katello stuff for now. I will take a look to add it to katello.

**Comment by bagasse** [comment]
Katello issue: [#27545](https://projects.theforeman.org/issues/27545)

**Comment by bagasse** [comment]
Ok... :man_facepalming: katello already support lifecycle environments via API... My bad... but it's named `environments` and not `lifecycle_environments`.

**Comment by dabelenda** [comment]
What about `activation_key` ? the only way to setup the `activation_key` is through the `host_collections` parameter of the `activation_key` but AFAIK that will change the activation_key of all hosts in the collection after they are added to the collection AND the playbook is run.

**Comment by bagasse** [comment]
@dabelenda, I think that activation key can be set via parameters via the `kt_activation_keys` string param.

**Comment by evgeni** [comment]
fixed in #389, IMHO this can be closed now?
